BayesianEstimation/AdaptiveMetropolisWithinGibbs.cs

```diff
@@ -0,0 +1,285 @@
+
+        /// <summary>
+        /// Sets up the posterior function.
+        /// </summary>
+        protected void SetUpPosteriorFunction(bool isTwoSample)
+        {
+            // initialize parameters
+            var pooled = Data.SelectMany(p => p.Select(v => v)).ToList();
+
+            // the initial parameters are set to:
+            // mu: the pooled data's mean
+            // sd: the pooled data's sd
+            // nu: 5
+            if (isTwoSample)
+            {
+                numParams = 5;
+                currentState = new double[] { pooled.Average(), pooled.Average(), pooled.StandardDeviation(), pooled.StandardDeviation(), 5 };
+            }
+            else
+            {
+                numParams = 3;
+                currentState = new double[] { pooled.Average(), pooled.StandardDeviation(), 5 };
+            }
+
+            logSd = new double[numParams];
+            acceptanceCount = new double[numParams];
+        }
+
+        /// <summary>
+        /// Calculates the probability that a set of parameters explains the data (the posterior probability).
+        /// </summary>
+        protected double GetPosteriorProbability(double[] parameters)
+        {
+            double[] mu = null;
+            double[] sigma = null;
+            double nu = 0;
+
+            if (isTwoSample)
+            {
+                mu = new double[] { parameters[0], parameters[1] };
+                sigma = new double[] { parameters[2], parameters[3] };
+                nu = parameters[4];
+            }
+            else
+            {
+                mu = new double[] { parameters[0] };
+                sigma = new double[] { parameters[1] };
+                nu = parameters[2];
+            }
+
+            double log_p = 0;
+            log_p += Math.Log(Exponential.PDF(exp, nu - 1.0)); // estimate nu
+
+            for (var sample = 0; sample < mu.Length; sample++) // estimate mu and sd for each sample
+            {
+                log_p += Math.Log(ContinuousUniform.PDF(sigmaLow, sigmaHigh, sigma[sample]));
+
+                double normalPriorProb = Normal.PDF(meanMu, sdMu, mu[sample]);
+                log_p += Math.Log(normalPriorProb);
+
+                for (var subj_i = 0; subj_i < Data[sample].Count; subj_i++)
+                {
+                    double mmt = Data[sample][subj_i];
+                    double studentTPdf = StudentT.PDF(mu[sample], sigma[sample], nu, mmt);
+                    log_p += Math.Log(studentTPdf);
+                }
+            }
+
+            return log_p;
+        }
+        
+        /// <summary>
+        /// Samples the parameter distributions via MCMC.
+        /// </summary>
+        private double[] next_sample()
+        {
+            markovChain.Add(currentState.ToArray());
+
+            for (var param_i = 0; param_i < numParams; param_i++)
+            {
+                var parameterProposal = Normal.Sample(random, currentState[param_i], Math.Exp(logSd[param_i]));
```

Comment:
is the assumption from everything that distributions are normal. maybe that should be stated at the top?